[FEATURE] 중고거래 상품 등록 기능 구현

u-genuine opened the pull request:
---
name: PULL_REQUEST_TEMPLATE
about: Describe this issue template's purpose here.
title: ''
labels: ''
assignees: ''

---

### PR Type
<!— Please check the one that applies to this PR using "x". —>

- [x] 버그수정(Bugfix)
- [x] 기능개발(Feature)
- [ ] 코드 스타일 변경(Code style update) (formatting, local variables)
- [x] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경
- [ ] 문서 내용 변경
- [ ] Other… Please describe:

### 요약(Summary)
중고거래 상품 등록 기능 구현 
도메인 엔티티/JPA 엔티티 분리
유효성 검증 강화
테스트 코드 작성

서버 DB 테이블 변경사항 반영
- product
- product_category
- product_photo

### 상세 내용(Describe your changes)

#### 엔티티 생성
- Product
- ProductPhoto
- ProductCategory
- ProductCategoryMapping

#### 레포지토리 인터페이스화

#### 이미지 유효성 검증 강화
- ImageValidator에 MIME 타입 체크 추가

#### 테스트 코드 작성
- TestUserHelper 클래스 추가
- ProductService, ProductAppender, ProductPhotoAppender, ProductCategoryFinder 테스트 추가

#### 중고거래 상품 등록 기능 구현
  
<img width="933" alt="image" src="https://github.com/user-attachments/assets/338f38a4-e579-45bc-a6da-18d57ca7b443" />

**응답 예시**
<img width="902" alt="image" src="https://github.com/user-attachments/assets/38aa2327-7c79-4b27-8a58-740fb0bed213" />
<img width="891" alt="image" src="https://github.com/user-attachments/assets/57f459fd-0da4-45c2-a304-076b0d890ce2" />
<img width="893" alt="image" src="https://github.com/user-attachments/assets/e1340841-225e-4063-ae19-eeccdd340f98" />
<img width="901" alt="image" src="https://github.com/user-attachments/assets/51595dbb-06a0-4a18-a68c-2ba3871bf05a" />
<img width="900" alt="image" src="https://github.com/user-attachments/assets/c3c186c3-921a-4c79-b67b-83a0d35eb5ed" />
<img width="895" alt="image" src="https://github.com/user-attachments/assets/4e44e2b7-ecf4-404b-92e6-a6860030a3b6" />
<img width="893" alt="image" src="https://github.com/user-attachments/assets/c100c3de-aa5d-422f-a983-8bdff6e61a96" />
<img width="888" alt="image" src="https://github.com/user-attachments/assets/5a3b3dad-fb05-4df6-a736-4e733dbdd388" />
<img width="891" alt="image" src="https://github.com/user-attachments/assets/1ac17107-c531-4083-acef-2b9190f777fd" />
<img width="886" alt="image" src="https://github.com/user-attachments/assets/764a8f9a-f33c-45f2-be3a-eeade45605bd" />

### Issue Number or Link
Resolve #209

u-genuine committed:
refactor(IT-Cotato#209): Product 도메인 엔티티, JPA 엔티티 분리 & 유효성 검증 로직 추가

u-genuine committed:
refactor(IT-Cotato#209): ProductPhoto 도메인 엔티티, JPA 엔티티 분리 & 유효성 검증 로직 추가

u-genuine committed:
refactor(IT-Cotato#209): Product, ProductPhoto 레포지토리 인터페이스화

- xxxRepository: 비즈니스 계층에서 사용할 인터페이스
- xxxRepositoryImpl: 실제 구현체
- xxxJpaRepository: Spring Data JPA 인터페이스

u-genuine committed:
feature(IT-Cotato#209): ProductCategory 엔티티 생성

    - ProductCategory: 중고거래에 사용되는 카테고리
    - ProductCategoryMapping 상품과 카테고리 매핑 엔티티

u-genuine committed:
feature(IT-Cotato#209): ProductCategory, ProductCategoryMapping 레포지토리 인터페이스화

u-genuine committed:
feature(IT-Cotato#209): ImageValidator 이미지 유효성 검증 강화, MIME 타입 체크 추가

- 이미지 업로드 최대 개수(10개) 제한 추가
- 허용된 MIME 타입 검증 로직 구현 (jpeg, png, gif, webp)
- 상세한 에러 메시지 구현 (지원되지 않는 형식, 빈 파일 등)

u-genuine committed:
feature(IT-Cotato#209): 중고거래 상품 등록 기능 구현

u-genuine committed:
fix(IT-Cotato#209): 에러코드 오타 수정

u-genuine committed:
feat(IT-Cotato#209): Product 유효성 검증 추가

- 가격이 음수일 떄 예외처리

u-genuine committed:
test(IT-Cotato#209): ProductAppender 테스트 코드 작성

- 상품 등록 성공 케이스
- 상품 등록 실패: userId null
- 상품 등록 실패: price null
- 상품 등록 실패: price 음수
- 상품 등록 실패: title 빈 문자열
- 상품 등록 실패: description 빈 문자열

u-genuine committed:
bugfix(IT-Cotato#209): ProductPhotoAppender가 ProductPhotoRepository를 의존하도록 수정

u-genuine committed:
bugfix(IT-Cotato#209): ImageValidator 메서드 분리

- validateProductImages: 중고거래 필수 이미지 검증용
- validateImagesCommon: 공통 검증 로직

u-genuine committed:
test(IT-Cotato#209): ProductPhotoAppender 테스트 코드 작성

- 상품 사진 목록을 순서대로 저장

u-genuine committed:
test(IT-Cotato#209): ProductCategoryFinder 테스트 코드 작성

- 존재하는 카테고리 이름으로 조회 성공
- 존재하지 않는 카테고리 이름으로 조회 시 예외 발생

u-genuine committed:
bugfix(IT-Cotato#209): 상품 등록 시 카테고리 필수 검증 추가

u-genuine committed:
test(IT-Cotato#209): 상품 카테고리 매핑 테스트 코드 작성

- 상품에 여러 카테고리 매핑, 저장 성공

u-genuine committed:
test(IT-Cotato#209): TestUserHelper 클래스 추가

- UserDto 테스트용 객체 생성을 위한 팩토리 클래스
- id, universityId, userRole만 커스텀하여 생성 가능

u-genuine committed:
test(IT-Cotato#209): ProductService 테스트 코드 작성

- 상품 생성 성공: 모든 파라미터가 유효한 경우
- 상품 생성 실패: 재학 인증되지 않은 유저
- 상품 생성 실패: 카테고리가 없는 경우
- 상품 생성 실패: 카테고리가 유효하지 않은 경우
- 상품 생성 실패: 유효하지 않은 이미지

u-genuine committed:
bugfix(IT-Cotato#209): photo_order로 컬럼명 변경하여 MySQL 예약어 충돌 해결

u-genuine committed:
docs(IT-Cotato#209): 중고거래 API Swagger 문서화 추가

u-genuine committed:
refactor(IT-Cotato#209): 이지기 유효성 검증 에러 개선

u-genuine committed:
bugfix(IT-Cotato#209): status 필드 설정 누락 수정

goalSetter09 commented on src/main/java/com/cotato/kampus/domain/product/domain/Product.java:
```diff
@@ -1,74 +1,171 @@
-	@Enumerated(EnumType.STRING)
-	@Column(name = "category", nullable = false)
-	private ProductCategory productCategory;
+	private void validate() {
```
컨트롤러에서 서비스로 넘어올 때, 검증이 한 번 이루어지는데 또 검증하는 게 필요한가요?!
다른 용도가 있는건가요??

u-genuine replied:
컨트롤러에서만 요청이 들어오는 게 아니라 테스트 코드 같은 다른 경로를 통해서도 도메인 객체를 생성할 수 있어서 도메인 엔티티 자체에서 유효성 검증이 있어야 할 것 같아요!

goalSetter09 replied:
음... 다른 이유가 아니라 테스트 코드 때문에 도메인의 코드가 수정되어야 하는거면, 고민이 필요할 것 같아요! private 필드 값을 테스트 하기 위해서 굳이 setter를 만들지 않듯이요..

정답은 없겠지만, 저는 타입이나 값의 유효성을 판단하는 로직은 request 단에서 처리를 하고, 도메인 객체에서는 비즈니스 로직과 관련된 검증만 있으면 좋지 않을까 싶습니다!

관련 레퍼런스 자료 첨부합니다!
https://www.inflearn.com/community/questions/1320209/%EA%B0%9D%EC%B2%B4-%EC%83%9D%EC%84%B1-%EC%8B%9C-%EA%B2%80%EC%A6%9D-%EB%A1%9C%EC%A7%81%EC%97%90-%EB%8C%80%ED%95%B4-%EC%A7%88%EB%AC%B8%EB%93%9C%EB%A6%BD%EB%8B%88%EB%8B%A4?focusComment=353413

u-genuine replied:
말씀하신 대로 다시 생각해보니 값의 필수 여부 같은 기본적인 유효성 검증은 requestDTO에서 처리하는 것이 더 적절한 것 같습니다!

개발자가 객체 생성할 때 필드를 누락하는 실수를 방지하기 위해 도메인에 검증 로직을 추가했었는데 검증 비용까지 생각하면 비즈니스 로직 관련 검증만 하는게 나은 것 같아요!

수정하겠습니당 꼼꼼한 리뷰 감사해요🥹

goalSetter09 replied:
아닙니당! 저도 고민을 하던 주제였는데 좀 명확하게 정리할 수 있는 것 같아요!
정리를 하자면,
1. 객체 타입, 값의 유효성 등의 기본적인 검사: request dto 단에서 검증
2. 비즈니스 로직과 관련된 검증: domain 객체
3. 비즈니스 로직과 관련되어 있지만 객체 하나로 검증이 불가능한 경우: validator와 같은 클래스를 생성하여 검증
예시)
- 중복된 객체 검증 시에는 db에서 꺼내봐야 알기 때문에 validator가 필요
- 자기 자신에게 채팅하는 걸 검증하는 경우: domain 객체 내에서 검증

이렇게 규칙을 정해서 사용하면 좋을 것 같습니다!

u-genuine replied:
수정된 부분 확인 부탁드립니다! 😊

goalSetter09 commented on src/main/java/com/cotato/kampus/domain/product/implement/product/ProductAppender.java:
```diff
@@ -0,0 +1,32 @@
+	@Transactional
+	public Product append(Long userId, String title, Integer price, String description) {
+
+		Product product = Product.builder()
```
엇 검증 로직이 여기서는 사용이 안되네요??

u-genuine replied:
객체 생성할 때 Product 클래스 생성자 안에서 validate() 메서드가 항상 실행되고 있습니다!

goalSetter09 commented on src/main/java/com/cotato/kampus/domain/product/domain/Product.java:
```diff
@@ -1,74 +1,171 @@
+		this.status = status;
+		this.createdTime = createdTime;
+		this.lastModifiedTime = lastModifiedTime;
+		validate();
```
검증 로직이 밑에있어서, 검증에서 실패하는 로직이더라도 객체가 생성은 될 것 같아용
가장 위로 올리면 좋을 것 같습니다!

그리고 public으로 열려있어서 유효성 검증이 스킵될 수 있는 위험이 있어보입니다!

u-genuine replied:
클래스에는 @Getter만 붙이고 @Builder는 생성자에 따로 붙어있어서 생성자가 호출될때마다 validate()가 호출됩니다! 
만약 누락된 필드가 있으면 예외를 던져서 객체가 생성되지 않더라구용

goalSetter09 replied:
아하 제 말은 검증 로직이 마지막에 실행되면 각 필드가 값 할당이 이루어진 후 검증 로직에서 예외를 던지기 때문에 크게 의미는 없겠지만 위로 올리는게 더 효율적이지 않을까? 라는 의미였습니다!! 물론 결과는 같지만용

> 클래스에는 @Getter만 붙이고 @builder는 생성자에 따로 붙어있어서 생성자가 호출될때마다 validate()가 호출됩니다!

이거에 대한 거는 예를 들어 지금 withUpdateInfo()를 사용해서 업데이트를 하는 것으로 보이는데, 이걸 사용하는 걸 강제할 수 없기 때문에 나중에 실수가 나올 수 있다는 말이었습니다!

u-genuine replied:
아 이제 이해됐네요ㅜㅜ
withUpdateInfo()를 사용하지 않고 빌더로 업데이트 시도할 경우에 필드가 잘못 초기화될 수도 있고 잘못된 값이 할당될 수도 있네요..!!
생성자를 private으로 바꾸고 팩토리 메서드로 수정하겠습니다!!

goalSetter09 reviewed (commented):
수고하셨습니다!

u-genuine committed:
refactor(IT-Cotato#209): Product 생성자 private으로 변경하고 팩토리 메서드 추가

- 생성자 접근 제어자를 private으로 변경하여 생성 방식 제한

- 도메인 객체 생성 시 create/fromEntity 팩토리 메서드 사용하도록 변경

- 필드 유효성 검사는 생성자에서 제거하고 외부(Request)에서만 처리하도록 설계 변경

u-genuine committed:
rename(IT-Cotato#209): RequestDTO 이름 수정

u-genuine committed:
test(IT-Cotato#209): 필요 없는 테스트 코드 제거

u-genuine committed:
refactor(IT-Cotato#209): 필수값 유효성 검증 로직 제거
